lib/modules/wdpa/dopa_importer.rb

```diff
@@ -0,0 +1,32 @@
+module Wdpa::DopaImporter
+  DOPA_LIST = "#{Rails.root}/lib/data/seeds/dopa4_pas_list.csv"
+  extend self
+
+  def import
+    ActiveRecord::Base.transaction do
+      # Clear all previously set DOPA areas
+      ProtectedArea.update_all(is_dopa: false)
+      logger = Logger.new(STDOUT)
+
+      pas_to_update = []
+
+      # Read the CSV to get the WDPA IDs
+      CSV.foreach(DOPA_LIST, headers: true) do |row| 
+        # Find the associated Protected Area with each WDPA ID mentioned and update the is_dopa column to true
+        dopa_pa = ProtectedArea.where('reported_area > 0.5e1')
+                  .or(ProtectedArea.where('reported_marine_area > 0.5e1'))
+                  .find_by_wdpa_id(row['wdpaid'])
+
+        if dopa_pa.nil? 
+          logger.info "DOPA site #{row['wdpaid']} not present in DB"
+          next
+        end
+
+       pas_to_update << ProtectedArea.find_by_wdpa_id(row['wdpaid'])
+      end
+      
+      ProtectedArea.where(id: pas_to_update.map(&:id)).update_all(is_dopa: true)
```

Comment:
A couple of considerations here:

1. If you want to make an early check on the existance of the PA, it would better to just save the WDPA ID of the PA found instead of saving the entire PA object and doing a where statement getting the ids again.
2. Further to that, to do way less requests to the DB, you could first get all WDPA IDs from the CSV file and then use the `where` statement to fetch the found PAs. SQL will manage those that can be found automatically.
3. If you are going for the approach above, you could also make a diff between the WDPA IDs in the CSV file and areas found so to then print a statement of the number (and even the related WDPA IDs) of the not found PAs. You could achieve this with something like:

```
  wdpa_ids_from_csv = # read WDPA IDs from CSV
  areas_found = ProtectedArea.where(id: wdpa_ids_from_csv)
  ids_not_found = wdpa_ids_from_csv - areas_found.map(&:wdpa_id)
```

Reply:
That's a fantastic idea, also TIL that you can pass an array of IDs to the `where` method.